python script update for order

bonidier opened the pull request:
Hi @bittner,

Thanks for your work !

Your Python script is very useful !

I think Lego has changed their website since your last script commit : the order URL in your script is now only to notify Lego about missing part of a product already buy.

some other problems : 
- the quantity selection can't work as expected.
- selenium 2.48.0 works only on recent Chrome version but hang with Firefox.

So I've made a lot of change, updated requirements.txt about Selenium version.

I've also added a note about required browser's drivers in the README file.

Here is a test of this new script with Selenium/browser compare : 

(only tested with Linux)

**selenium 2.48.0**
- chrome 60.0.3112.113
   - no login : OK
   - login : OK
- firefox 55.0.3 : hang, not compatible

**selenium 3.5.0**
- chrome 60.0.3112.113 
   - no login : OK
   - login : OK
- firefox 55.0.3 
   - no login : OK
   - login : OK

Regards

bonidier committed:
python script update :
- fix Lego shop url
- add Chrome option to keep Chrome opened with latest Selenium version
- remove stuff about sold-out-info alert, N/A with new url
- fix selection of item's amount
- maximize browser window to avoid element detection error
- rework login authentication, detect failed login
- rework detection of elements
- add counters : remaining product to add/not in set/sold-out)
- doc : update requirements

bittner commented on README.rst:
```diff
@@ -77,6 +77,11 @@ Usage
+
+   * Python Selenium (see requirements.txt)
+   * `geckodriver`_
+   * `chromedriver`_
```
Thanks for your PR!

Can you please move your additions into a new section "Requirements" (right before "Documentation, ...", use a "~~~~" underlining as with the other sections).

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -174,11 +178,38 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+
+    shop_url = 'https://wwwsecure.us.lego.com/{shop}/service/replacementparts/sale'.format(shop=shop)
+
+    print("Browser URL : ",shop_url)
```
I'm not sure this will work across Python 2 and 3 versions. Please use string formatting instead, e.g.
``` python
print("Browser URL: {url}".format(url=shop_url))
```

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -174,11 +178,38 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+    ## detect browser choice
+    if browser == 'chrome':
+        opts = ChromeOptions()
+        # With selenium version above this one, chrome is closed at the end without the "quit()" method !
```
Please wrap the line before 80 characters are reached, and don't use a space before the exclamation mark.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -190,26 +221,42 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+
         print("Let's log in with LEGO ID {user}.".format(user=username))
-        login_link = browser.find_element_by_css_selector('.legoid .links > a')
+        login_link = wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,".legoid .links > a[data-uitest = 'login-link']")))
```
- Please wrap the line before 80 characters line length.
- Please use a space after the comma.
- Please don't use spaces around the `=` sign in the CSS selector.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -190,26 +221,42 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
         browser.switch_to.frame('legoid-iframe')
 
-        user_input = browser.find_element_by_id('fieldUsername')
+        user_input = wait.until(EC.element_to_be_clickable((By.ID,'fieldUsername')))
```
Space after the comma, please.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -190,26 +221,42 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
         user_input.send_keys(username)
-        passwd_input = browser.find_element_by_id('fieldPassword')
+
+        passwd_input = wait.until(EC.element_to_be_clickable((By.ID,'fieldPassword')))
```
Same here, space after the comma, please.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -190,26 +221,42 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
-    wait = WebDriverWait(browser, 5)
+        # ensure the user/password are good
+        try:
+          wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,".legoid .links > a[data-uitest = 'logout-link']")))
```
Same as above:

- Please wrap the line before 80 characters line length.
- Please use a space after the comma.
- Please don't use spaces around the `=` sign in the CSS selector.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -190,26 +221,42 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+          wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR,".legoid .links > a[data-uitest = 'logout-link']")))
+          print("login success !")
+        except TimeoutException:
+          print("login failed !")
```
No (french-style) space before punctuations, please. Both `print`s above.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -221,49 +268,67 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+    counter_notInSet = 0
+    counter_found = 0
+
+    total_elements=len(order_list)
```
PEP8 wants us to use spaces around the `=` sign here.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -221,49 +268,67 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
 
+        counter += 1
+
+        print("- [{counter}/{total_elements}] {qty}x #{pn} ".format(qty=quantity, pn=part_no,counter=counter,total_elements=total_elements), end='')
```
Please wrap this before 80 characters, too.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -221,49 +268,67 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+              sleep(.2)  # seconds
+            else:
+              counter_outOfStock += 1
+              print("NOTE: item out of stock. ")
```
Is the space at the end of the sentence needed?

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -221,49 +268,67 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+
         except NoSuchElementException:
             print("OOOPS! No LEGO part with that number found in set #{set}. :-(".format(set=lego_set))
+            counter_notInSet += 1
```
No camel case, please! Should better be `not_in_set_counter` (and probably `found_counter` above).

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -221,49 +268,67 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+    print("Stats :")
+    print("Elements found          : {s}/{total_elements}".format(s=counter_found,total_elements=total_elements))
+    print("Elements not in set     : {s}".format(s=counter_notInSet))
+    print("Elements out of stock   : {s}".format(s=counter_outOfStock))
```
In general, no spaces before punctuations in English. And, what would you think if we got rid of the aligning problem by placing the numbers in front, hmm? Would that be okay? Kind of ...
```
Statistics:
- xx elements found.
- yy elements not in set.
- zz elements out of stock.
```
No worries about alignment then.

bittner reviewed (changes requested):
Good additions, thanks! Please run `flake8` over the whole source code. That always helps keeping the code clean a little. Cheers! -- _Merci beaucoup !_

bonidier committed:
python script review
- pass on order() with flake8 linter
- Selenium : fix detection of rpAgeAndCountryAgeField, may fail sometimes
- rework statistics output
- README : add dedicated section about python script requirements

bonidier commented:
Hi @bittner , I've fixed all your requests and more

flake8 report no error, except some "E501 line too long" due to printed strings
I've done my best to reduce the number of E501 in order() function

tested with : 
```
flake8 --ignore=E501 lego-mindstorms-pieces.py
```

Regards !

bittner commented:
Thanks a lot for your contribution, @bonidier! Merci beaucoup !
